### PR TITLE
Adapt AAPP reader for generic chunk size

### PR DESCRIPTION
**Pull request by mraspaud**
This PR allows passing chunksize like "128MiB" to the aapp reader.

### CODE REVIEW

**Review by djhoese — COMMENTED** [review]
Well...I guess I never finished my review of this. GitHub says my comments are still pending. Sorry. I made these comments a long long time ago.

File: satpy/readers/aapp_l1b.py [hunk 1]
```diff
@@ -299,10 +314,11 @@ def calibrate(self,
 
         if dataset_id['name'] in ("3a", "3b") and self._is3b is None:
             # Is it 3a or 3b:
+            line_chunks = get_chunks((self._data.shape[0], 2048))[0]
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:317]: as the variable name hints, I looking for the number of lines per chunks. `get_chunks` returns line, col chunks sizes

[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/readers/aapp_l1b.py:317]: My question was more why are you doing this rather than what are you doing. But I see now that these are scan-based fields, right? So they are one dimensional along-track variables (the row dimension) so there is no need to specify column-wise chunks.

[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:317]: yes indeed, sorry for the confusion

File: satpy/readers/aapp_l1b.py [hunk 2]
```diff
@@ -36,7 +36,22 @@
 from satpy import CHUNK_SIZE
 from satpy.readers.file_handlers import BaseFileHandler
 
-LINE_CHUNK = CHUNK_SIZE ** 2 // 2048
+CHANNEL_DTYPE = float
+
+
+def get_chunks(shape):
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:42]: yes, we could make this a generic function, but maybe we should leave the normalize_chunks out of it so that each reader can just call `normalize_chunks` with it's preferred parameters (position of "auto", dtype). So this function would just be actually `get_limit_from_pytroll_chunk_size` or something, that can be removed when we deprecated the environment variable.

`normalize_chunks` does a good size at keeping the user preferences and splitting things evenly, for example if we think of AHI:
```
In [5]: da.core.normalize_chunks((2200, "auto"), shape=(22000, 22000), dtype='float')
Out[5]: 
((2200, 2200, 2200, 2200, 2200, 2200, 2200, 2200, 2200, 2200),
 (5500, 5500, 5500, 5500))
```

So, in summary, I think we should leave `normalize_chunks` to the individual readers, and have them call this function for getting the limit.

[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/readers/aapp_l1b.py:42]: Some complications/questions with this idea:

1. `CHANNEL_DTYPE` is not an argument to this function, it should be.
2. Couldn't `("auto", 2048)` be an argument to the function too?

If the above is done then I don't see a reason it couldn't stay with the same functionality. This way also the function could be updated with a very descriptive/detailed docstring to help new reader developers understand the decisions that need to be made when choosing chunks for a reader. Additionally, having a central function for controlling things as important as chunking might be nice.

We may need to think about how something like resampling might use this though. For example, the output chunks of resampling based on a target AreaDefinition. That could be handled by this method as well. Or can you think of a reason that it would need a separate `satpy.config` configuration option on top of dask's chunk size config?

[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:42]: That's the thing, at the moment PYTROLL_CHUNK_SIZE know nothing of prefered chunks size or dtype, so I think that keeping the same behaviour (just counting the number of pixels) through having this `get_limits` function is good.
After that, I don't see why we shouldn't use `normalize_chunks` directly, that does basically what we need, and the reader dev just has to look at that documentation to understand how to get the chunks sizes they think is best for that specific reader. Or am i missing something?

[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/readers/aapp_l1b.py:42]: Ah for some reason I read the `if` statement cases backwards and thought the `else:` was when they specified a string. Right now PYTROLL_CHUNK_SIZE is used in a lot of places (even in pyresample I think), so it will be a long time before we can completely deprecated it I think.

I was going to say what if we just completely removed it one reader at a time and had a deprecation warning saying that PYTROLL_CHUNK_SIZE isn't having an effect. But I'm not sure DeprecatingWarning is supposed to be used after the feature is completely removed. Plus that warning could go in this limit function. So I support your idea of moving the limit calculation to a helper function with an added deprecation warning (which don't show up by default). In the future I think we can make it a FutureWarning or even a UserWarning when all uses are removed.

That said, is there any reason to support `PYTROLL_CHUNK_SIZE` being a string in this function? It won't actually work until all use cases of PYTROLL_CHUNK_SIZE are updated or removed. Maybe just a warning pointing people to the dask config setting documentation?

[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:42]: Indeed, if we plan on deprecating the use of CHUNK_SIZE entirely, then we don't need to check if it is a string. What I thought when I read this is that we would evolve CHUNK_SIZE to accept the same arguments as `normalize_chunks`, but maybe deprecating this entirely is best.

[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:42]: So in summary, I can change this function to just return a limit and put it in `satpy.utils`

File: satpy/readers/aapp_l1b.py [hunk 3]
```diff
@@ -33,10 +33,22 @@
 import xarray as xr
 from dask import delayed
 
-from satpy import CHUNK_SIZE
 from satpy.readers.file_handlers import BaseFileHandler
+from satpy.utils import get_chunk_pixel_size
+
+CHANNEL_DTYPE = float
```
Comment (djhoese) [inline satpy/readers/aapp_l1b.py:39]: Why `float` versus the more explicit `np.float64`? Mostly just curious.

Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:39]: My understanding was that using plain python type was prefered, but I could be wrong :)

File: satpy/utils.py [hunk 4]
```diff
@@ -485,3 +487,17 @@ def ignore_invalid_float_warnings():
...
         warnings.simplefilter("ignore", RuntimeWarning)
         yield
+
+
+def get_chunk_pixel_size():
+    """Compute the maximum chunk size in bites from CHUNK_SIZE."""
+    if CHUNK_SIZE is None:
+        return None
+    elif isinstance(CHUNK_SIZE, str):
+        raise NotImplementedError
```
Comment (djhoese) [inline satpy/utils.py:497]: Will this ever be implemented?

File: satpy/readers/aapp_l1b.py [hunk 5]
```diff
@@ -33,10 +33,22 @@
...
+
+
+def get_avhrr_lac_chunks(shape, dtype):
+    """Get chunks from a given shape adapted for full-resolution AVHRR data."""
+    limit = get_chunk_pixel_size() * np.dtype(dtype).itemsize
```
Comment (djhoese) [inline satpy/readers/aapp_l1b.py:44]: Why can't the dtype logic be put in the `get_chunk_pixel_size` function?

**Review by djhoese — COMMENTED** [review]
A couple questions, but looks pretty good so far

File: satpy/readers/aapp_l1b.py (hunk 3 above)
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:39]: I can't see anything about my previous comment anywhere, so I'll just make this explicit.

Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:39]:
```suggestion
CHANNEL_DTYPE = np.float64
```

Comment (djhoese) [reply satpy/readers/aapp_l1b.py]: Using the plain python type is preferred over the deprecated `np.float`, but I'm not sure `float` is preferred over other numpy types.

File: satpy/readers/aapp_l1b.py [hunk 6]
```diff
@@ -33,10 +33,22 @@
...
+    return da.core.normalize_chunks(("auto", 2048), shape=shape, limit=limit, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:45]: I thought it would be more readable to have the standard lac data width here, but I can change it if you think it's better with -1

[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/readers/aapp_l1b.py:45]: Not a huge deal. I just thought it was more flexible to say `-1` to make it explicit that we want the whole width of the scan/swath.

[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:45]: I'll leave it like this then :)

File: satpy/readers/aapp_l1b.py (hunk 5 above)
Comment (mraspaud) [reply satpy/readers/aapp_l1b.py:44]: yes, it can, but then we have to change the name of that function. On it.

File: satpy/utils.py (hunk 4 above)
Comment (mraspaud) [reply satpy/utils.py:497]: Probably not. I'll remove this

File: satpy/readers/aapp_l1b.py (hunk 5 above)
Comment (djhoese) [reply satpy/readers/aapp_l1b.py:44]: Well is `pixel_size` even the right descriptor here? It isn't the chunk size of a pixel? It is the chunk size in pixels, right? So `get_chunk_size_in_pixels`? Or maybe more generically in not say `pixels` at all since chunk size is always based on elements (pixels).

So for dtype it would be `get_chunk_size_for_dtype`?

File: satpy/readers/hrpt.py [hunk 7]
```diff
@@ -137,6 +137,11 @@ def times(self):
         """Get the timestamps for each line."""
         return time_seconds(self._data["timecode"], self.year)
 
+    @cached_property
+    def chunks(self):
```
Comment (djhoese) [inline satpy/readers/hrpt.py:141]: Thoughts about prefixing with a `_`?

**Review by djhoese — APPROVED** [review]
Small suggestion to prefix with `_` but not required. So :+1: from me.

**Review by djhoese — APPROVED** [review]
LGTM. Feel free to merge.
